Performance improvements & Ruby 1.8.7 compatibility

mudge opened the pull request:
Attempt to speed up matching by dropping `Object#tap` for a local variable and other tricks.

By doing this (and switching hash syntax), we gain compatibility with Ruby 1.8.7.

mudge committed:
Avoid tap in favour of a local variable

Object#tap has been 22% slower than using a local variable in some
benchmarks we've run on other internal projects [0] (presumably due to
the cost of instantiating a block). Drop it in favour of explicitly
using a local variable.

[0]: https://github.com/altmetric/til/blob/master/ruby/benchmarking.md

mudge committed:
Use the reader to access Node attributes

Rather than accessing instance variables inside Node, use the getters
already provided by attr_reader. On a similar note, switch :suffix to be
an attr_accessor as it has both a reader and a writer.

mudge committed:
Use the truthiness of parent rather than nil?

Save a method call by using the truthiness of parent itself rather than
calling Object#nil? on it.

mattmacleod commented:
Benchmark suggests it's slightly faster without `#tap` :+1:

``` ruby
Benchmark.bmbm do |x|
  x.report('#tap') { AhoCorasickMatcherWithTap.new(File.readlines('/usr/share/dict/words').map(&:chomp)).match(File.read('/usr/share/dict/words')) }
  x.report('without #tap') { AhoCorasickMatcherWithoutTap.new(File.readlines('/usr/share/dict/words').map(&:chomp)).match(File.read('/usr/share/dict/words')) }
end

                  user     system      total        real
#tap          7.520000   0.080000   7.600000 (  7.728031)
without #tap  7.160000   0.070000   7.230000 (  7.226881)
```

mudge committed:
Add Hound configuration

By default, Hound uses a different Rubocop configuration [0] than we
have configured for this project. Instead, tell it to use the bundled
configuration.

[0]: https://raw.githubusercontent.com/thoughtbot/hound/master/config/style_guides/ruby.yml

mudge committed:
Add compatibility with Ruby 1.8.7

Specifically, use the pre-1.9 hash syntax and enforce it with Rubocop.

mudge committed:
Build against Rubinius 2 to fix build error

Building against rbx-2.4 gives the following error on Travis CI:

> Gemset '' does not exist, 'rvm rbx-2.4 do rvm gemset create ' first,
> or append '--create'.

Avoid this by using a less specific Rubinius version.

mudge committed:
Minor README tweaks

mudge committed:
Remove explicit require_paths as it's the default

By default, RubyGems will define `require_paths` to be `lib` [0] so we
can drop this from the gemspec.

[0]: http://guides.rubygems.org/specification-reference/#require_paths=
